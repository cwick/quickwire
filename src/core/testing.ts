import Platform from "qw/core/platform.ts";
import { fn } from "qw/platform/testing.ts";

// Not in the mood to write a custom testing framework. Let the platform handle it.
export * from "qw/platform/testing.ts";

export function mockPlatform(): Platform {
  return {
    log: fn() as () => void,
    serve: fn() as () => void,
    import: fn(() => Promise.resolve({})) as () => Promise<object>,
    projectDir: "test_project",
    exit: fn() as () => void,
  };
}
