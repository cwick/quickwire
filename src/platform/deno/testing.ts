export { describe, it } from "deno-std/testing/bdd.ts";
export { expect, fn } from "deno-std/expect/mod.ts";
export { assertSnapshot } from "deno-std/testing/snapshot.ts";
