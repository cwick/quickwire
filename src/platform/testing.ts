export * from "./deno/testing.ts";
