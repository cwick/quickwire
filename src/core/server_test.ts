import { describe, it, mockPlatform, expect, fn } from "qw/core/testing.ts";
import Server from "qw/core/server.ts";
import { assertSnapshot } from "deno-std/testing/snapshot.ts";

describe("Server", () => {
  it("serves", async (t) => {
    const platform = mockPlatform();
    platform.import = fn(() =>
      Promise.resolve({ default: () => "hello" })
    ) as () => Promise<object>;

    const server = new Server(platform);
    await assertSnapshot(t, await request(server, "/"));
  });

  it("serves different paths", async (t) => {
    const platform = mockPlatform();
    platform.import = fn((path: string) => {
      if (path === "routes/index.tsx") {
        return Promise.resolve({ default: () => "index" });
      } else if (path === "routes/about.tsx") {
        return Promise.resolve({ default: () => "about" });
      }
    }) as () => Promise<object>;

    const server = new Server(platform);
    await assertSnapshot(t, await request(server, "/"));
    await assertSnapshot(t, await request(server, "/about"));
  });

  async function request(server: Server, path: string) {
    const response = await server.handleRequest(
      new Request(`http://localhost:8080${path}`)
    );
    expect(response.status).toEqual(200);
    return await response.text();
  }
});
