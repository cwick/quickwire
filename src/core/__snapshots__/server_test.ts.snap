export const snapshot = {};

snapshot[`Server > serves 1`] = `'<html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><title></title></head><body>hello</body></html>'`;

snapshot[`Server > serves different paths 1`] = `'<html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><title></title></head><body>index</body></html>'`;

snapshot[`Server > serves different paths 2`] = `'<html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><title></title></head><body>about</body></html>'`;
